app/lib/coursesHelper.coffee

```diff
@@ -0,0 +1,192 @@
+      
+      for course, courseIndex in courses.models
+        campaign = campaigns.get(course.get('campaignID'))
+        
+        for level in campaign.getNonladderLevels().models
+          levelID = level.get('original')
+          
+          for concept in level.get('concepts')
+            if concept == 'basic_syntax'
+              console.log('level: ', level.attributes.campaign, 'concept: ', concept)
+            unless conceptData[classroom.id][concept]
+              conceptData[classroom.id][concept] = { completed: true, started: false }
+
+          for concept in level.get('concepts')
+            for userID in classroom.get('members')
+              session = _.find classroom.sessions.models, (session) ->
+                session.get('creator') is userID and session.get('level').original is levelID
+              
+              if not session # haven't gotten to this level yet, but might have completed others before
+                for concept in level.get('concepts')
+                  conceptData[classroom.id][concept].completed = false
+              if session # have gotten to the level and at least started it
+                for concept in level.get('concepts')
+                  conceptData[classroom.id][concept].started = true
+              if not session?.completed() # level started but not completed
+                for concept in level.get('concepts')
+                  conceptData[classroom.id][concept].completed = false
+    conceptData
+      
+  calculateAllProgress: (classrooms, courses, campaigns, courseInstances, students) ->
+    # Loop through all combinations and record:
+    #   Completeness for each student/course
+    #   Completeness for each student/level
+    #   Completeness for each class/course (across all students)
+    #   Completeness for each class/level (across all students)
+    
+    # class -> course
+    #   class -> course -> student
+    #   class -> course -> level
+    #     class -> course -> level -> student
+    
+    progressData = {}
+    for classroom in classrooms.models
+      progressData[classroom.id] = {}
+
+      for course, courseIndex in courses.models
+        instance = courseInstances.findWhere({ courseID: course.id, classroomID: classroom.id })
+        if not instance
+          progressData[classroom.id][course.id] = { completed: false, started: false }
+          continue
+        progressData[classroom.id][course.id] = { completed: true, started: false } # to be updated
+        
+        campaign = campaigns.get(course.get('campaignID'))
+        for level in campaign.getNonladderLevels().models
+          levelID = level.get('original')
+          progressData[classroom.id][course.id][levelID] = { completed: true, started: false }
+          
+          for user in students.models
+            userID = user.id
+            progressData[classroom.id][course.id][userID] ?= { completed: true, started: false } # Only set it the first time through a user
```

Comment:
`courseProgress = progressData[classroom.id][course.id]` and then use that throughout the rest of this block, to avoid so much repetition and long lines?